---
title: "Nucleotide diversity at different methylation levels and standard deviations"
author: "James Ord"
date: "4/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# PACKAGES
library(dplyr)
library(tidyr)
library(gridExtra)
library(matrixStats)
library(forcats)
library(OneR)

# FUNCTIONS

`%not_in%` = Negate(`%in%`)

# writeGTF_SiteClass() takes a three-column dataframe as input.
# the columns must be (1) chromosome, (2) site coordinate, and (3) site class
writeGTF_SiteClass <- function(input,output) {
  sitedata<-input
  sitedata$source<-"none"
  sitedata$feature<-"transcript"
  sitedata$score<-"."
  sitedata$strand<-"+"
  sitedata$frame<-"."
  sitedata$gene_id<-paste("\"",sitedata[,3],"\"",sep="")
  sitedata$attribute<-paste("gene_id ",sitedata$gene_id,";",sep="")
  GTF<-data.frame(seqname=sitedata[,1],
                source=sitedata$source,
                feature=sitedata$feature,
                start=sitedata[,2],
                end=sitedata[,2],
                score=sitedata$score,
                strand=sitedata$strand,
                frame=sitedata$frame,
                attribute=sitedata$attribute)
  write.table(GTF,file=output,quote=FALSE,col.names=FALSE,row.names=FALSE,sep="\t")
}
```
Read in our favourite data file
```{r}
sitesNsamples<-read.table(gzfile("DMC_data_subsample2_140422.tsv.gz"),header=T)
```

Make bins of the mean and SD of methylation in each result category for each population. We need two gtfs for this because we want to look at a different set of sites in each population
```{r}
# BINS OF MEAN METHYLATION PER RESULT PER POPULATION
for (i in c("Marine","Freshwater")){
meth_levels_long_pop<-subset(meth_levels_long,Population==i)

meth_mean_bins_pop<-NULL
for (j in c("hyper","hypo","non")){
meth_levels_long_result<-subset(meth_levels_long_pop,result_pop==j)
meth_bins<-bin(meth_levels_long_result,method="content", nbins = 50)
meth_bins$mean_bin<-as.numeric(meth_bins$mean)
meth_mean_bins<-merge(meth_bins[c(6,8)],meth_levels_long_result,by="ID")
head(meth_mean_bins)

mean_binmeth<-meth_mean_bins %>%
   group_by(mean_bin) %>%
   summarise(binmean_mean = mean(mean),
             bin_n=n())

meth_mean_bins<-merge(meth_mean_bins,mean_binmeth,by="mean_bin")
meth_mean_bins$Result_mean_n<-paste(meth_mean_bins$result_pop,meth_mean_bins$binmean_mean,meth_mean_bins$bin_n,sep="_")
meth_mean_bins_pop<-rbind(meth_mean_bins_pop,meth_mean_bins)
}
pop_acc<-ifelse(i=="Marine","SRR7470095","SRR869609") # write the population accession into the filename
writeGTF_SiteClass(meth_mean_bins_pop[c(3,4,11)],
                      output=paste("C:/Users/jms_o/Documents/R/pop_mean_bins",pop_acc,"150422.gtf",sep="_"))
}

# BINS OF SD OF METHYLATION PER RESULT PER POPULATION
for (i in c("Marine","Freshwater")){
meth_levels_long_pop<-subset(meth_levels_long,Population==i)

meth_SD_bins_pop<-NULL
for (j in c("hyper","hypo","non")){
meth_levels_long_result<-subset(meth_levels_long_pop,result_pop==j)
meth_bins<-bin(meth_levels_long_result,method="content", nbins = 50)
meth_bins$SD_bin<-as.numeric(meth_bins$SD)
meth_SD_bins<-merge(meth_bins[c(6,8)],meth_levels_long_result,by="ID")
head(meth_SD_bins)

SD_binmeth<-meth_SD_bins %>%
   group_by(SD_bin) %>%
   summarise(binmean_SD = mean(SD),
             bin_n=n())

meth_SD_bins<-merge(meth_SD_bins,SD_binmeth,by="SD_bin")
meth_SD_bins$Result_SD_n<-paste(meth_SD_bins$result_pop,meth_SD_bins$binmean_SD,meth_SD_bins$bin_n,sep="_")
meth_SD_bins_pop<-rbind(meth_SD_bins_pop,meth_SD_bins)
}
pop_acc<-ifelse(i=="Marine","SRR7470095","SRR869609") # write the population accession into the filename
writeGTF_SiteClass(meth_SD_bins_pop[c(3,4,11)],
                      output=paste("C:/Users/jms_o/Documents/R/pop_SD_bins",pop_acc,"150422.gtf",sep="_"))
}
```

Then with those GTFs, we can get the nucleotide diversity with popoolation, as usual. Well, not quite as per usual. This time we have a triple loop. The top layer works through the the two bin categories (mean and SD bins), then the second and third work through the populations and diversity measures, respectively. There are four GTFs: one for each combination of bin category and population.

```{r, engine = 'bash', eval = FALSE}
#!/bin/bash
#SBATCH --mail-type=all
#SBATCH --mail-user=james.ord@vetsuisse.unibe.ch
#SBATCH --job-name="TK_popoolation_mean_SD_bins_15422.sh"
#SBATCH --time=10:00:00
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=100G

for j in mean SD; do

for i in SRR869609:10 SRR7470095:12; do
i1=$(echo $i | cut -d ":" -f 1)
i2=$(echo $i | cut -d ":" -f 2)

for k in pi theta D; do

perl /storage/homefs/jo20n766/software/popoolation_1.2.2/Variance-at-position.pl --pool-size ${i2} --min-qual 20 --min-coverage 3 --min-count 2 --fastq-type sanger \
--pileup /storage/workspaces/vetsuisse_fiwi_epev/fiwi_epev001/datasets/dataset_collection/stickleback/PRJNA204958_479509/mapped/MAPQ20_rdup/pileup/${i1}.pileup \
--gtf pop_${j}_bins_${i1}_150422.gtf --output ${i1}_temp.${k} --measure ${k}

sed -i "s/$/\t${i1}/" ${i1}_temp.${k}

done

paste ${i1}_temp.pi <(cut -f 4 ${i1}_temp.theta) <(cut -f 4 ${i1}_temp.D) > ${i1}_${j}_temp.div

done

cat *_${j}_temp.div > pop_${j}_bins_150422.div
rm *_temp.*

done

```
CHANGE IN MEAN / SD

Bins of change in standard deviation:
```{r}
head(sitesNsamples)

meth_shifts<-sitesNsamples
meth_shifts$SD_shift<-meth_shifts$SD_FatF-meth_shifts$SD_MatM

meth_shifts$result_pop <- factor(meth_shifts$result_pop,levels=c("non","hypo","hyper"))

methlevels_hist<-ggplot(meth_shifts,aes(x=SD_shift,fill=result_pop))+
  theme_bw()+
  geom_histogram(alpha=0.5,position="identity",color="black")+
  labs(x="Shift in SD of methylation",y="Count")+
  scale_fill_manual(values=c("gray33","royalblue3","darkorange3"))

head(meth_shifts)

# BINS OF MEAN METHYLATION PER RESULT PER POPULATION
meth_shifts_for_gtf<-NULL
for (j in c("hyper","hypo","non")){
meth_shifts_result<-subset(meth_shifts,result_pop==j)

meth_bins<-bin(meth_shifts_result,method="content", nbins = 50)

meth_bins$SDS_bin<-as.numeric(meth_bins$SD_shift)
meth_mean_bins<-merge(meth_bins[c(13,15)],meth_shifts_result,by="ID")

mean_binmeth<-meth_mean_bins %>%
   group_by(SDS_bin) %>%
   summarise(binSDS_mean = mean(SD_shift),
             bin_n=n())

meth_mean_bins<-merge(meth_mean_bins,mean_binmeth,by="SDS_bin")
meth_mean_bins$Result_mean_n<-paste(meth_mean_bins$result_pop,meth_mean_bins$binSDS_mean,meth_mean_bins$bin_n,sep="_")
meth_shifts_for_gtf<-rbind(meth_shifts_for_gtf,meth_mean_bins)
}

writeGTF_SiteClass(meth_shifts_for_gtf[c(3,4,18)],
                      output=paste("C:/Users/jms_o/Documents/R/pop_SDS_bins_190422.gtf",sep="_"))
```

Next, bins of the change in mean % methylation (basically extent of differential methylation)
```{r}
meth_shifts_for_gtf<-NULL
for (j in c("hyper","hypo")){
meth_shifts_result<-subset(meth_shifts,result_pop==j)

meth_bins<-bin(meth_shifts_result,method="content", nbins = 50)

meth_bins$methdiff_bin<-as.numeric(meth_bins$meth_diff_pop)
meth_mean_bins<-merge(meth_bins[c(13,15)],meth_shifts_result,by="ID")

mean_binmeth<-meth_mean_bins %>%
   group_by(methdiff_bin) %>%
   summarise(bin_methdiff_mean = mean(meth_diff_pop),
             bin_n=n())

meth_mean_bins<-merge(meth_mean_bins,mean_binmeth,by="methdiff_bin")
meth_mean_bins$Result_mean_n<-paste(meth_mean_bins$result_pop,meth_mean_bins$bin_methdiff_mean,meth_mean_bins$bin_n,sep="_")
meth_shifts_for_gtf<-rbind(meth_shifts_for_gtf,meth_mean_bins)
}

writeGTF_SiteClass(meth_shifts_for_gtf[c(3,4,18)],
                      output=paste("C:/Users/jms_o/Documents/R/pop_methdiff_bins_190422.gtf",sep="_"))

```