---
title: "DMR_calling_271121"
author: "James Ord"
date: "11/27/2021 | edited 05/04/22"
output: html_document
---

```{r setup, include=FALSE}
library(bsseq)
library(GenomicRanges)
library(methylSig)
```
```{r}
# read in the files: here we want files 1-3 and 9-11 which comprise the population comparison.
testAT <- read.bismark(list.files("/cov",
                                  full.names = TRUE)[c(1:3,9:11)],
                       colData=data.frame(t=c("M","M","M","F","F","F")))

# Filtering out unwanted chromosomes:
testgrAT <- granges(testAT)
testgrAT_rm <- testgrAT[(seqnames(testgrAT)=="chrXIX")|(seqnames(testgrAT)=="chrM")|(seqnames(testgrAT)=="chrY")]
testAT<-filter_loci_by_location(testAT,testgrAT_rm)
testgrAT<-NULL;testgrAT_rm<-NULL
```

```{r}
test.fitAT <- BSmooth(
    BSseq = testAT, 
    verbose = TRUE)

#?getCoverage
BS.covAT <- getCoverage(test.fitAT) # coverage is PER BASE

keepLoci.ex <- which(rowSums(BS.covAT[, test.fitAT$t == "M"] >= 5) >= 3 &
                     rowSums(BS.covAT[, test.fitAT$t == "F"] >= 5) >= 3)

length(test.fitAT)
length(keepLoci.ex)

BS.ex.fitAT <- test.fitAT[keepLoci.ex,]
```

```{r}
pData(BS.ex.fitAT)

BS.ex.tstatAT <- BSmooth.tstat(BS.ex.fitAT,
                                    group1 = c(1,2,3),
                                    group2 = c(4,5,6),
                                    estimate.var = "same",
                                    local.correct = TRUE,
                                    verbose = TRUE)


```
See the distribution of T-values. Should be basically normal.
```{r}
BS.ex.tstatAT
plot(BS.ex.tstatAT)
```
Here we set reasonable T-stat, number of Cs, and mean methylation difference to consider a region differentially methylated.
Here, a region is differentially methylated if the absolute value of the T-stat is >=1.5, there are at least 5 Cs and the mean methylation difference is at least 15%.
```{r}
dmrs0AT <- dmrFinder(BS.ex.tstatAT, cutoff = c(-1.5, 1.5))

dmrsAT <- subset(dmrs0AT, n >= 5 & abs(meanDiff) >= 0.15)

# by default the 'hyper' and 'hypo' labels are assigned in terms of group 1 vs group 2. I want them the other way around.
dmrsAT$direction<-ifelse(dmrsAT$direction=="hyper","hypo","hyper")
nrow(subset(dmrsAT,direction=="hyper"))
nrow(subset(dmrsAT,direction=="hypo"))

head(dmrsAT)
```
Let's now make a BED file for each comparison.
```{r}
write.table(subset(dmrsAT,direction=="hyper")[1:3],file="DMRs_AT_hyper_271121.bed",row.names = FALSE,col.names = FALSE,quote=FALSE,sep="\t")
write.table(subset(dmrsAT,direction=="hypo")[1:3],file="DMRs_AT_hypo_271121.bed",row.names = FALSE,col.names = FALSE,quote=FALSE,sep="\t")
```
